File: test_go/src/pattern/Combining/duck/eg6_observer/flock.go
package eg6_observer

//群（实现Quackable接口）
//组合需要和叶节点元素实现相同的接口；这里的叶节点就是Quackable
type Flock struct {
	quackers []Quackable //这里最好使用带有迭代器的容器
}

func NewFlock() *Flock {
	this := new(Flock)
	this.quackers = make([]Quackable, 0, 10)

	return this
}

func (this *Flock) Add(quacker Quackable) {
	this.quackers = append(this.quackers, quacker)
}

func (this *Flock) Quack() {
	for _, quacker := range this.quackers {
		if v, ok := quacker.(*Flock); ok {
			v.Quack()
		} else {
			quacker.Quack()
		}
	}
}

func (this *Flock) RegisterObserver(observer Observer) {
	for _, quacker := range this.quackers {
		if v, ok := quacker.(*Flock); ok {
			v.RegisterObserver(observer)
		} else {
			quacker.RegisterObserver(observer)
		}
	}
}

func (this *Flock) NotifyObserver() {
}

func (this *Flock) String() string {
	return "Flock of Quackers"
}

File: java_test/src/DesignPattern/Combining/Duck/Eg6Observer/Flock.java
package DesignPattern.Combining.Duck.Eg6Observer;

import java.util.ArrayList;
import java.util.Iterator;

public class Flock implements Quackable {
	ArrayList<Quackable> ducks = new ArrayList<Quackable>();
 
	public void add(Quackable quacker) {
		ducks.add(quacker);
	}
 
	public void quack() {
		Iterator<Quackable> iterator = ducks.iterator();
		while (iterator.hasNext()) {
			Quackable quacker = iterator.next();
			quacker.quack();
		}
	}

	public void registerObserver(Observer observer) {
		Iterator<Quackable> iterator = ducks.iterator();
		while (iterator.hasNext()) {
			Quackable duck = (Quackable)iterator.next();
			duck.registerObserver(observer);
		}
	}

	public void notifyObservers() { }
	public String toString() {
		return "Flock of Quackers";
	}
}
